File: swift/Sources/MLXAudioModels/HTDemucs/Transformer.swift
// Transformer.swift
// Transformer layers for HTDemucs.

import Foundation
@preconcurrency import MLX
import MLXNN

// MARK: - Positional Embeddings

/// Create 1D sinusoidal positional embedding.
///
/// Matches PyTorch demucs create_sin_embedding exactly.
/// - Parameters:
///   - length: Sequence length
///   - dim: Embedding dimension
///   - shift: Position shift (default 0)
///   - maxPeriod: Maximum period for sinusoidal encoding
/// - Returns: Positional embedding `[length, 1, dim]`
public func createSinEmbedding(
    length: Int,
    dim: Int,
    shift: Int = 0,
    maxPeriod: Float = 10000.0
) -> MLXArray {
    // pos: [T, 1, 1]
    let positions = MLXArray((shift..<(shift + length)).map { Float($0) })
    let pos = positions.expandedDimensions(axes: [1, 2])

    let halfDim = dim / 2
    let adim = MLXArray(0..<halfDim).asType(.float32).expandedDimensions(axes: [0, 1])

    // phase = pos / (maxPeriod ** (adim / (halfDim - 1)))
    let exponent = adim / Float(halfDim - 1)
    let divisor = pow(MLXArray(maxPeriod), exponent)
    let phase = pos / divisor

    let cosPhase = cos(phase)
    let sinPhase = sin(phase)
    return concatenated([cosPhase, sinPhase], axis: -1)  // [T, 1, D]
}

/// Create 2D sinusoidal positional embedding.
///
/// Matches PyTorch demucs create_2d_sin_embedding exactly:
/// - First half of channels: encodes width with sin/cos interleaved
/// - Second half of channels: encodes height with sin/cos interleaved
///
/// - Parameters:
///   - dModel: Embedding dimension (must be divisible by 4)
///   - height: Height (freq bins)
///   - width: Width (time frames)
///   - maxPeriod: Maximum period
/// - Returns: Positional embedding `[1, dModel, height, width]`
public func create2DSinEmbedding(
    dModel: Int,
    height: Int,
    width: Int,
    maxPeriod: Float = 10000.0
) -> MLXArray {
    let halfD = dModel / 2

    // div_term: step by 2 to match PyTorch
    let indices = MLXArray(stride(from: 0, to: halfD, by: 2).map { Float($0) })
    let divTerm = exp(indices * Float(-log(Double(maxPeriod)) / Double(halfD)))  // [D/4]

    let posW = MLXArray(0..<width).asType(.float32).expandedDimensions(axis: 1)  // [W, 1]
    let posH = MLXArray(0..<height).asType(.float32).expandedDimensions(axis: 1)  // [H, 1]

    // Width encoding
    let wArgs = posW * divTerm  // [W, D/4]
    let sinW = sin(wArgs)
    let cosW = cos(wArgs)

    // Height encoding
    let hArgs = posH * divTerm  // [H, D/4]
    let sinH = sin(hArgs)
    let cosH = cos(hArgs)

    // Interleave sin/cos: [W, D/4] -> stack -> [W, D/4, 2] -> reshape [W, D/2]
    var widthEmb = stacked([sinW, cosW], axis: -1)  // [W, D/4, 2]
    widthEmb = widthEmb.reshaped([width, halfD])  // [W, D/2]
    widthEmb = widthEmb.T  // [D/2, W]
    // Broadcast over height: [D/2, W] -> [D/2, H, W]
    widthEmb = widthEmb.expandedDimensions(axis: 1)
    widthEmb = MLX.broadcast(widthEmb, to: [halfD, height, width])

    var heightEmb = stacked([sinH, cosH], axis: -1)  // [H, D/4, 2]
    heightEmb = heightEmb.reshaped([height, halfD])  // [H, D/2]
    heightEmb = heightEmb.T  // [D/2, H]
    // Broadcast over width: [D/2, H] -> [D/2, H, W]
    heightEmb = heightEmb.expandedDimensions(axis: 2)
    heightEmb = MLX.broadcast(heightEmb, to: [halfD, height, width])

    let pe = concatenated([widthEmb, heightEmb], axis: 0)  // [D, H, W]
    return pe.expandedDimensions(axis: 0)  // [1, D, H, W]
}

// MARK: - MultiheadAttention

/// Multihead attention matching PyTorch's nn.MultiheadAttention.
///
/// Uses combined in_proj for Q, K, V (in_proj_weight, in_proj_bias).
public class MultiheadAttention: Module, @unchecked Sendable {
    let embedDim: Int
    let numHeads: Int
    let headDim: Int
    let scale: Float

    @ParameterInfo(key: "in_proj_weight") var in_proj_weight: MLXArray
    @ParameterInfo(key: "in_proj_bias") var in_proj_bias: MLXArray
    let out_proj: Linear

    /// Creates a MultiheadAttention layer.
    /// - Parameters:
    ///   - embedDim: Embedding dimension.
    ///   - numHeads: Number of attention heads.
    ///   - dropout: Dropout rate (unused, kept for API compatibility).
    public init(embedDim: Int, numHeads: Int, dropout: Float = 0.0) {
        self.embedDim = embedDim
        self.numHeads = numHeads
        self.headDim = embedDim / numHeads
        self.scale = pow(Float(headDim), -0.5)

        // Combined QKV projection (MLX format: [in, out])
        // PyTorch [3*D, D] -> transposed to MLX [D, 3*D]
        self._in_proj_weight.wrappedValue = MLXArray.zeros([embedDim, 3 * embedDim])
        self._in_proj_bias.wrappedValue = MLXArray.zeros([3 * embedDim])

        // Output projection
        self.out_proj = Linear(embedDim, embedDim)
    }

    /// Forward pass.
    /// - Parameters:
    ///   - query: Query tensor `[B, T, D]`
    ///   - key: Key tensor `[B, S, D]`
    ///   - value: Value tensor `[B, S, D]`
    /// - Returns: Output tensor `[B, T, D]`
    public func callAsFunction(
        query: MLXArray,
        key: MLXArray,
        value: MLXArray
    ) -> MLXArray {
        let B = query.shape[0]
        let T = query.shape[1]
        let S = key.shape[1]

        // Project Q, K, V using combined weight
        let wQ = in_proj_weight[0..., 0..<embedDim]
        let wK = in_proj_weight[0..., embedDim..<(2 * embedDim)]
        let wV = in_proj_weight[0..., (2 * embedDim)...]

        let bQ = in_proj_bias[0..<embedDim]
        let bK = in_proj_bias[embedDim..<(2 * embedDim)]
        let bV = in_proj_bias[(2 * embedDim)...]

        var q = matmul(query, wQ) + bQ
        var k = matmul(key, wK) + bK
        var v = matmul(value, wV) + bV

        // Reshape for multi-head attention
        q = q.reshaped([B, T, numHeads, headDim]).transposed(0, 2, 1, 3)
        k = k.reshaped([B, S, numHeads, headDim]).transposed(0, 2, 1, 3)
        v = v.reshaped([B, S, numHeads, headDim]).transposed(0, 2, 1, 3)

        // Attention
        var attn = matmul(q, k.transposed(0, 1, 3, 2)) * scale
        attn = softmax(attn, axis: -1)

        // Output
        var out = matmul(attn, v)
        out = out.transposed(0, 2, 1, 3).reshaped([B, T, embedDim])

        return out_proj(out)
    }
}

// MARK: - MyTransformerEncoderLayer

/// Transformer encoder layer matching PyTorch's MyTransformerEncoderLayer.
///
/// Structure (self-attention variant):
/// - self_attn: MultiheadAttention
/// - linear1, linear2: FFN
/// - norm1, norm2: LayerNorm
/// - norm_out: MyGroupNorm
/// - gamma_1, gamma_2: LayerScale
///
/// Structure (cross-attention variant, adds):
/// - cross_attn: MultiheadAttention
/// - norm3: LayerNorm
public class MyTransformerEncoderLayer: Module, @unchecked Sendable {
    let crossAttention: Bool

    var self_attn: MultiheadAttention?
    var cross_attn: MultiheadAttention?

    let linear1: Linear
    let linear2: Linear

    let norm1: LayerNorm
    let norm2: LayerNorm
    var norm3: LayerNorm?

    let norm_out: MyGroupNorm
    let gamma_1: LayerScale
    let gamma_2: LayerScale

    /// Creates a transformer encoder layer.
    /// - Parameters:
    ///   - dModel: Model dimension.
    ///   - nhead: Number of attention heads.
    ///   - dimFeedforward: FFN hidden dimension.
    ///   - dropout: Dropout rate (unused).
    ///   - crossAttention: If true, use cross-attention instead of self-attention.
    public init(
        dModel: Int,
        nhead: Int,
        dimFeedforward: Int,
        dropout: Float = 0.0,
        crossAttention: Bool = false
    ) {
        self.crossAttention = crossAttention

        if crossAttention {
            self.cross_attn = MultiheadAttention(embedDim: dModel, numHeads: nhead, dropout: dropout)
            self.self_attn = nil
        } else {
            self.self_attn = MultiheadAttention(embedDim: dModel, numHeads: nhead, dropout: dropout)
            self.cross_attn = nil
        }

        self.linear1 = Linear(dModel, dimFeedforward)
        self.linear2 = Linear(dimFeedforward, dModel)

        self.norm1 = LayerNorm(dimensions: dModel)
        self.norm2 = LayerNorm(dimensions: dModel)
        if crossAttention {
            self.norm3 = LayerNorm(dimensions: dModel)
        }

        self.norm_out = MyGroupNorm(numChannels: dModel)
        self.gamma_1 = LayerScale(channels: dModel, init: 1e-4)
        self.gamma_2 = LayerScale(channels: dModel, init: 1e-4)
    }

    /// Forward pass.
    /// - Parameters:
    ///   - x: Input `[B, T, D]`
    ///   - cross: Cross-attention source `[B, S, D]` (only for cross-attn layers)
    /// - Returns: Output `[B, T, D]`
    public func callAsFunction(_ x: MLXArray, cross: MLXArray? = nil) -> MLXArray {
        var out = x

        if crossAttention {
            // Cross-attention
            let xNorm = norm1(out)
            let crossNorm = norm2(cross!)
            let attnOut = cross_attn!(query: xNorm, key: crossNorm, value: crossNorm)
            out = out + gamma_1(attnOut)

            // FFN
            let xNorm2 = norm3!(out)
            let ffnOut = linear2(gelu(linear1(xNorm2)))
            out = out + gamma_2(ffnOut)
        } else {
            // Self-attention
            let xNorm = norm1(out)
            let attnOut = self_attn!(query: xNorm, key: xNorm, value: xNorm)
            out = out + gamma_1(attnOut)

            // FFN
            let xNorm2 = norm2(out)
            let ffnOut = linear2(gelu(linear1(xNorm2)))
            out = out + gamma_2(ffnOut)
        }

        // Final norm
        out = norm_out(out)

        return out
    }
}

// MARK: - CrossTransformerEncoder

/// Cross-domain transformer matching PyTorch's CrossTransformerEncoder.
///
/// Structure:
/// - norm_in: LayerNorm (freq input)
/// - norm_in_t: LayerNorm (time input)
/// - layers: list of MyTransformerEncoderLayer (freq branch)
/// - layers_t: list of MyTransformerEncoderLayer (time branch)
///
/// Pattern: alternating self-attention (0,2,4) and cross-attention (1,3)
public class CrossTransformerEncoder: Module, @unchecked Sendable {
    let dim: Int
    let depth: Int
    let maxPeriod: Float

    let norm_in: LayerNorm
    let norm_in_t: LayerNorm

    let layers: [MyTransformerEncoderLayer]
    let layers_t: [MyTransformerEncoderLayer]

    /// Creates a cross-domain transformer.
    /// - Parameters:
    ///   - dim: Model dimension.
    ///   - depth: Number of layers.
    ///   - heads: Number of attention heads.
    ///   - dimFeedforward: FFN hidden dimension (nil = 4*dim).
    ///   - dropout: Dropout rate.
    ///   - maxPeriod: Maximum period for positional embeddings.
    public init(
        dim: Int,
        depth: Int = 5,
        heads: Int = 8,
        dimFeedforward: Int? = nil,
        dropout: Float = 0.0,
        maxPeriod: Float = 10000.0
    ) {
        self.dim = dim
        self.depth = depth
        self.maxPeriod = maxPeriod

        let feedforward = dimFeedforward ?? (4 * dim)

        self.norm_in = LayerNorm(dimensions: dim)
        self.norm_in_t = LayerNorm(dimensions: dim)

        // Build layers - alternating self and cross attention
        var freqLayers: [MyTransformerEncoderLayer] = []
        var timeLayers: [MyTransformerEncoderLayer] = []

        for i in 0..<depth {
            let cross = (i % 2 == 1)  // odd layers are cross-attention
            freqLayers.append(
                MyTransformerEncoderLayer(
                    dModel: dim,
                    nhead: heads,
                    dimFeedforward: feedforward,
                    dropout: dropout,
                    crossAttention: cross
                )
            )
            timeLayers.append(
                MyTransformerEncoderLayer(
                    dModel: dim,
                    nhead: heads,
                    dimFeedforward: feedforward,
                    dropout: dropout,
                    crossAttention: cross
                )
            )
        }

        self.layers = freqLayers
        self.layers_t = timeLayers
    }

    /// Apply cross-domain transformer.
    /// - Parameters:
    ///   - freq: Frequency features `[B, C, F, T]` (4D)
    ///   - time: Time features `[B, C, T]` (3D)
    /// - Returns: Tuple of (freq_out, time_out)
    public func callAsFunction(
        freq: MLXArray,
        time: MLXArray
    ) -> (freq: MLXArray, time: MLXArray) {
        let B = freq.shape[0]
        let C = freq.shape[1]
        let Fr = freq.shape[2]
        let T1 = freq.shape[3]
        let T2 = time.shape[2]

        // Create 2D positional embedding for freq branch
        var posEmb2D = create2DSinEmbedding(dModel: C, height: Fr, width: T1, maxPeriod: maxPeriod)

        // Rearrange: [B, C, Fr, T1] -> [B, (T1*Fr), C]
        posEmb2D = posEmb2D.transposed(0, 3, 2, 1)  // [1, T1, Fr, C]
        posEmb2D = posEmb2D.reshaped([1, T1 * Fr, C])  // [1, T1*Fr, C]

        var freqFlat = freq.transposed(0, 3, 2, 1)  // [B, T1, Fr, C]
        freqFlat = freqFlat.reshaped([B, T1 * Fr, C])  // [B, T1*Fr, C]

        // Apply norm and add position embedding
        freqFlat = norm_in(freqFlat)
        freqFlat = freqFlat + posEmb2D

        // Create 1D positional embedding for time branch
        var posEmb1D = createSinEmbedding(length: T2, dim: C, shift: 0, maxPeriod: maxPeriod)
        posEmb1D = posEmb1D.transposed(1, 0, 2)  // [1, T2, C]

        // Rearrange: [B, C, T2] -> [B, T2, C]
        var timeFlat = time.transposed(0, 2, 1)

        // Apply norm and add position embedding
        timeFlat = norm_in_t(timeFlat)
        timeFlat = timeFlat + posEmb1D

        // Apply layers
        for i in 0..<depth {
            if i % 2 == 0 {
                // Self-attention
                freqFlat = layers[i](freqFlat)
                timeFlat = layers_t[i](timeFlat)
            } else {
                // Cross-attention (freq attends to time, time attends to freq)
                let oldFreq = freqFlat
                freqFlat = layers[i](freqFlat, cross: timeFlat)
                timeFlat = layers_t[i](timeFlat, cross: oldFreq)
            }
        }

        // Rearrange back: [B, (T1*Fr), C] -> [B, C, Fr, T1]
        var freqOut = freqFlat.reshaped([B, T1, Fr, C])
        freqOut = freqOut.transposed(0, 3, 2, 1)

        // Rearrange back: [B, T2, C] -> [B, C, T2]
        let timeOut = timeFlat.transposed(0, 2, 1)

        return (freqOut, timeOut)
    }
}

File: python/mlx_audio/models/demucs/layers/transformer.py
"""Transformer layers for HTDemucs.

Matches PyTorch demucs.transformer exactly.
"""

from __future__ import annotations

import math

import mlx.core as mx
import mlx.nn as nn

from mlx_audio.models.demucs.layers.dconv import LayerScale


def create_sin_embedding(
    length: int, dim: int, shift: int = 0, max_period: float = 10000.0
) -> mx.array:
    """Create 1D sinusoidal positional embedding.

    Matches PyTorch demucs create_sin_embedding exactly.
    Returns TBC format: [length, 1, dim].

    Args:
        length: Sequence length
        dim: Embedding dimension
        shift: Position shift (default 0)
        max_period: Maximum period for sinusoidal encoding

    Returns:
        Positional embedding [length, 1, dim]
    """
    pos = (shift + mx.arange(length, dtype=mx.float32))[:, None, None]  # [T, 1, 1]
    half_dim = dim // 2
    adim = mx.arange(half_dim, dtype=mx.float32)[None, None, :]  # [1, 1, D/2]
    phase = pos / (max_period ** (adim / (half_dim - 1)))
    emb = mx.concatenate([mx.cos(phase), mx.sin(phase)], axis=-1)  # [T, 1, D]
    return emb


def create_2d_sin_embedding(
    d_model: int, height: int, width: int, max_period: float = 10000.0
) -> mx.array:
    """Create 2D sinusoidal positional embedding.

    Matches PyTorch demucs create_2d_sin_embedding exactly:
    - First half of channels: encodes width with sin/cos interleaved
    - Second half of channels: encodes height with sin/cos interleaved

    Args:
        d_model: Embedding dimension (must be divisible by 4)
        height: Height (freq bins)
        width: Width (time frames)
        max_period: Maximum period

    Returns:
        Positional embedding [1, d_model, height, width]
    """
    half_d = d_model // 2
    quarter_d = half_d // 2

    # div_term: step by 2 to match PyTorch
    div_term = mx.exp(
        mx.arange(0.0, half_d, 2, dtype=mx.float32)
        * -(math.log(max_period) / half_d)
    )  # [D/4]

    pos_w = mx.arange(width, dtype=mx.float32)[:, None]  # [W, 1]
    pos_h = mx.arange(height, dtype=mx.float32)[:, None]  # [H, 1]

    # Width encoding
    w_args = pos_w * div_term[None, :]  # [W, D/4]
    sin_w = mx.sin(w_args)  # [W, D/4]
    cos_w = mx.cos(w_args)  # [W, D/4]

    # Height encoding
    h_args = pos_h * div_term[None, :]  # [H, D/4]
    sin_h = mx.sin(h_args)  # [H, D/4]
    cos_h = mx.cos(h_args)  # [H, D/4]

    # Interleave sin/cos: [W, D/4] -> stack -> [W, D/4, 2] -> reshape [W, D/2]
    width_emb = mx.stack([sin_w, cos_w], axis=-1)  # [W, D/4, 2]
    width_emb = width_emb.reshape(width, half_d)  # [W, D/2]
    width_emb = width_emb.T  # [D/2, W]
    # Broadcast over height: [D/2, W] -> [D/2, H, W]
    width_emb = width_emb[:, None, :]
    width_emb = mx.broadcast_to(width_emb, (half_d, height, width))

    height_emb = mx.stack([sin_h, cos_h], axis=-1)  # [H, D/4, 2]
    height_emb = height_emb.reshape(height, half_d)  # [H, D/2]
    height_emb = height_emb.T  # [D/2, H]
    # Broadcast over width: [D/2, H] -> [D/2, H, W]
    height_emb = height_emb[:, :, None]
    height_emb = mx.broadcast_to(height_emb, (half_d, height, width))

    pe = mx.concatenate([width_emb, height_emb], axis=0)  # [D, H, W]
    return pe[None, :, :, :]  # [1, D, H, W]


class MyGroupNorm(nn.Module):
    """GroupNorm that matches PyTorch's MyGroupNorm (1 group).

    PyTorch MyGroupNorm with groups=1:
    1. Transposes [B, T, C] -> [B, C, T]
    2. Normalizes over ALL of C and T together (not just C)
    3. Transposes back

    This is equivalent to normalizing over axes (1, 2) of [B, T, C].
    """

    def __init__(self, num_channels: int, eps: float = 1e-5):
        super().__init__()
        self.num_channels = num_channels
        self.weight = mx.ones((num_channels,))
        self.bias = mx.zeros((num_channels,))
        self.eps = eps

    def __call__(self, x: mx.array) -> mx.array:
        # x: [B, T, C]
        # GroupNorm(1, C) normalizes over all of T and C together
        mean = mx.mean(x, axis=(1, 2), keepdims=True)
        var = mx.var(x, axis=(1, 2), keepdims=True)
        x = (x - mean) / mx.sqrt(var + self.eps)
        return x * self.weight + self.bias


class MultiheadAttention(nn.Module):
    """Multihead attention matching PyTorch's nn.MultiheadAttention.

    Uses combined in_proj for Q, K, V (in_proj_weight, in_proj_bias).
    """

    def __init__(self, embed_dim: int, num_heads: int, dropout: float = 0.0):
        super().__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        self.head_dim = embed_dim // num_heads
        self.scale = self.head_dim ** -0.5

        # Combined QKV projection (MLX format: [in, out])
        # PyTorch [3*D, D] -> transposed to MLX [D, 3*D]
        self.in_proj_weight = mx.zeros((embed_dim, 3 * embed_dim))
        self.in_proj_bias = mx.zeros((3 * embed_dim,))

        # Output projection
        self.out_proj = nn.Linear(embed_dim, embed_dim)

    def __call__(
        self,
        query: mx.array,
        key: mx.array,
        value: mx.array,
    ) -> mx.array:
        """Forward pass.

        Args:
            query: [B, T, D]
            key: [B, S, D]
            value: [B, S, D]

        Returns:
            Output [B, T, D]
        """
        B, T, _ = query.shape
        S = key.shape[1]

        # Project Q, K, V using combined weight
        # in_proj_weight is [D, 3*D], split along axis 1
        w_q = self.in_proj_weight[:, :self.embed_dim]
        w_k = self.in_proj_weight[:, self.embed_dim:2*self.embed_dim]
        w_v = self.in_proj_weight[:, 2*self.embed_dim:]

        b_q = self.in_proj_bias[:self.embed_dim]
        b_k = self.in_proj_bias[self.embed_dim:2*self.embed_dim]
        b_v = self.in_proj_bias[2*self.embed_dim:]

        # query @ w_q: [B, T, D] @ [D, D] -> [B, T, D]
        q = query @ w_q + b_q
        k = key @ w_k + b_k
        v = value @ w_v + b_v

        # Reshape for multi-head attention
        q = q.reshape(B, T, self.num_heads, self.head_dim).transpose(0, 2, 1, 3)
        k = k.reshape(B, S, self.num_heads, self.head_dim).transpose(0, 2, 1, 3)
        v = v.reshape(B, S, self.num_heads, self.head_dim).transpose(0, 2, 1, 3)

        # Attention
        attn = (q @ k.transpose(0, 1, 3, 2)) * self.scale
        attn = mx.softmax(attn, axis=-1)

        # Output
        out = attn @ v
        out = out.transpose(0, 2, 1, 3).reshape(B, T, self.embed_dim)

        return self.out_proj(out)


class MyTransformerEncoderLayer(nn.Module):
    """Transformer encoder layer matching PyTorch's MyTransformerEncoderLayer.

    Structure (self-attention variant):
        self_attn: MultiheadAttention
        linear1: Linear
        dropout: (no params)
        linear2: Linear
        norm1: LayerNorm
        norm2: LayerNorm
        dropout1: (no params)
        dropout2: (no params)
        norm_out: MyGroupNorm
        gamma_1: LayerScale
        gamma_2: LayerScale

    Structure (cross-attention variant, adds):
        cross_attn: MultiheadAttention
        norm3: LayerNorm
    """

    def __init__(
        self,
        d_model: int,
        nhead: int,
        dim_feedforward: int,
        dropout: float = 0.0,
        cross_attention: bool = False,
    ):
        super().__init__()
        self.cross_attention = cross_attention

        if cross_attention:
            self.cross_attn = MultiheadAttention(d_model, nhead, dropout)
        else:
            self.self_attn = MultiheadAttention(d_model, nhead, dropout)

        self.linear1 = nn.Linear(d_model, dim_feedforward)
        self.linear2 = nn.Linear(dim_feedforward, d_model)

        self.norm1 = nn.LayerNorm(d_model)
        self.norm2 = nn.LayerNorm(d_model)
        if cross_attention:
            self.norm3 = nn.LayerNorm(d_model)

        self.norm_out = MyGroupNorm(d_model)
        self.gamma_1 = LayerScale(d_model, init=1e-4)
        self.gamma_2 = LayerScale(d_model, init=1e-4)

    def __call__(
        self,
        x: mx.array,
        cross: mx.array | None = None,
    ) -> mx.array:
        """Forward pass.

        Args:
            x: Input [B, T, D]
            cross: Cross-attention source [B, S, D] (only for cross-attn layers)

        Returns:
            Output [B, T, D]
        """
        if self.cross_attention:
            # Cross-attention
            x_norm = self.norm1(x)
            cross_norm = self.norm2(cross)
            attn_out = self.cross_attn(x_norm, cross_norm, cross_norm)
            x = x + self.gamma_1(attn_out)

            # FFN
            x_norm = self.norm3(x)
            ffn_out = self.linear2(nn.gelu(self.linear1(x_norm)))
            x = x + self.gamma_2(ffn_out)
        else:
            # Self-attention
            x_norm = self.norm1(x)
            attn_out = self.self_attn(x_norm, x_norm, x_norm)
            x = x + self.gamma_1(attn_out)

            # FFN
            x_norm = self.norm2(x)
            ffn_out = self.linear2(nn.gelu(self.linear1(x_norm)))
            x = x + self.gamma_2(ffn_out)

        # Final norm
        x = self.norm_out(x)

        return x


class CrossTransformerEncoder(nn.Module):
    """Cross-domain transformer matching PyTorch's CrossTransformerEncoder.

    Structure:
        norm_in: LayerNorm (freq input)
        norm_in_t: LayerNorm (time input)
        layers: list of MyTransformerEncoderLayer (freq branch)
        layers_t: list of MyTransformerEncoderLayer (time branch)

    Pattern: alternating self-attention (0,2,4) and cross-attention (1,3)

    Input format:
        freq: [B, C, F, T] (4D frequency features)
        time: [B, C, T] (3D time features)
    """

    def __init__(
        self,
        dim: int,
        depth: int = 5,
        heads: int = 8,
        dim_feedforward: int | None = None,
        dropout: float = 0.0,
        max_period: float = 10000.0,
    ):
        super().__init__()
        self.dim = dim
        self.depth = depth
        self.max_period = max_period

        # Position embedding weight (scalar, default 1.0)
        # This is not a learned parameter in pretrained model
        self.weight_pos_embed = 1.0

        if dim_feedforward is None:
            dim_feedforward = 4 * dim

        # Input layer norms
        self.norm_in = nn.LayerNorm(dim)
        self.norm_in_t = nn.LayerNorm(dim)

        # Build layers - alternating self and cross attention
        self.layers = []
        self.layers_t = []

        for i in range(depth):
            cross = (i % 2 == 1)  # odd layers are cross-attention
            self.layers.append(
                MyTransformerEncoderLayer(
                    dim, heads, dim_feedforward, dropout, cross_attention=cross
                )
            )
            self.layers_t.append(
                MyTransformerEncoderLayer(
                    dim, heads, dim_feedforward, dropout, cross_attention=cross
                )
            )

    def __call__(
        self,
        freq: mx.array,
        time: mx.array,
    ) -> tuple[mx.array, mx.array]:
        """Apply cross-domain transformer.

        Args:
            freq: Frequency features [B, C, F, T] (4D)
            time: Time features [B, C, T] (3D)

        Returns:
            Tuple of (freq_out, time_out):
                freq_out: [B, C, F, T]
                time_out: [B, C, T]
        """
        B, C, Fr, T1 = freq.shape
        _, _, T2 = time.shape

        # Create 2D positional embedding for freq branch
        pos_emb_2d = create_2d_sin_embedding(C, Fr, T1, self.max_period)

        # Rearrange: [B, C, Fr, T1] -> [B, (T1*Fr), C]
        # PyTorch does "b c fr t1 -> b (t1 fr) c" (time-major flattening)
        pos_emb_2d = pos_emb_2d.transpose(0, 3, 2, 1)  # [1, T1, Fr, C]
        pos_emb_2d = pos_emb_2d.reshape(1, T1 * Fr, C)  # [1, T1*Fr, C]

        freq = freq.transpose(0, 3, 2, 1)  # [B, T1, Fr, C]
        freq = freq.reshape(B, T1 * Fr, C)  # [B, T1*Fr, C]

        # Apply norm and add position embedding
        freq = self.norm_in(freq)
        freq = freq + self.weight_pos_embed * pos_emb_2d

        # Create 1D positional embedding for time branch
        # Returns [T2, 1, C] (TBC format), rearrange to [1, T2, C]
        pos_emb_1d = create_sin_embedding(T2, C, 0, self.max_period)  # [T2, 1, C]
        pos_emb_1d = pos_emb_1d.transpose(1, 0, 2)  # [1, T2, C]

        # Rearrange: [B, C, T2] -> [B, T2, C]
        time = time.transpose(0, 2, 1)

        # Apply norm and add position embedding
        time = self.norm_in_t(time)
        time = time + self.weight_pos_embed * pos_emb_1d

        # Apply layers
        for i, (freq_layer, time_layer) in enumerate(
            zip(self.layers, self.layers_t)
        ):
            if i % 2 == 0:
                # Self-attention
                freq = freq_layer(freq)
                time = time_layer(time)
            else:
                # Cross-attention (freq attends to time, time attends to freq)
                old_freq = freq
                freq = freq_layer(freq, time)
                time = time_layer(time, old_freq)

        # Rearrange back: [B, (T1*Fr), C] -> [B, C, Fr, T1]
        freq = freq.reshape(B, T1, Fr, C)
        freq = freq.transpose(0, 3, 2, 1)  # [B, C, Fr, T1]

        # Rearrange back: [B, T2, C] -> [B, C, T2]
        time = time.transpose(0, 2, 1)

        return freq, time
